File: Chapter05/Java/CryptoExample1/app/src/main/java/net/zenconsult/mobile/CryptoExample1Activity.java
package net.zenconsult.mobile;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;

import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class CryptoExample1Activity extends AppCompatActivity {

    private static final String TAG = "CryptoExample1Activity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    protected void onResume() {
        super.onResume();
        String key = "3106166021166038";
        String message = "This is a secret message";
        try {
            byte[] encryptData = encryptData(key, message);
            String decryptStr = decryptData(key, encryptData);
            Log.d(TAG, "onResume=>decryptStr: " + decryptStr);
        } catch (Exception e) {
            Log.e(TAG, "onResume=>error: ", e);
        }
    }

    private byte[] encryptData(String key, String data)
            throws NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidAlgorithmParameterException, InvalidKeyException,
            BadPaddingException, IllegalBlockSizeException {
        SecretKeySpec sks = new SecretKeySpec(key.getBytes(), "AES");
        Cipher c = Cipher.getInstance("AES/CBC/ZeroBytePadding");
        c.init(Cipher.ENCRYPT_MODE, sks, new IvParameterSpec(key.getBytes()));
        return c.doFinal(data.getBytes());
    }

    private String decryptData(String key, byte[] data)
            throws NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidAlgorithmParameterException, InvalidKeyException,
            BadPaddingException, IllegalBlockSizeException {
        SecretKeySpec sks = new SecretKeySpec(key.getBytes(), "AES");
        Cipher c = Cipher.getInstance("AES/CBC/ZeroBytePadding");
        c.init(Cipher.DECRYPT_MODE, sks, new IvParameterSpec(key.getBytes()));
        return new String(c.doFinal(data));
    }
}

File: Chapter05/Kotlin/CryptoExample1/app/src/main/java/net/zenconsult/mobile/CryptoExample1Activity.kt
package net.zenconsult.mobile

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class CryptoExample1Activity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onResume() {
        super.onResume()
        val key = "3106166021166038"
        val message = "This is a secret message"
        val encryptData = encryptData(key, message)
        val decryptStr = decryptData(key, encryptData)
        Log.d(TAG, "onResume=>decryptStr: $decryptStr")
    }

    fun encryptData(key: String, data: String): ByteArray {
        val sks = SecretKeySpec(key.toByteArray(), "AES")
        Log.d(TAG, "encryptData=>sks: ${sks.toString()}")
        val c = Cipher.getInstance("AES/CBC/ZeroBytePadding")
        c.init(Cipher.ENCRYPT_MODE, sks, IvParameterSpec(key.toByteArray()))
        return c.doFinal(data.toByteArray())
    }

    fun decryptData(key: String, data: ByteArray): String {
        val sks = SecretKeySpec(key.toByteArray(), "AES")
        Log.d(TAG, "decryptData=>sks: ${sks.toString()}")
        val c = Cipher.getInstance("AES/CBC/ZeroBytePadding")
        c.init(Cipher.DECRYPT_MODE, sks, IvParameterSpec(key.toByteArray()))
        return String(c.doFinal(data))
    }

    companion object {
        const val TAG = "CryptoExample1Activity"
    }
}